File: 005-longest palindromic substring/Solution005.cpp
#include <stdio.h>

//Attempt1: 108ms
class Solution005 {
public:
  string longestPalindrome(string s) {
    if (s.length() == 0)
      return "";

    string longest = s.substr(0, 1);
    for (int i = 0; i < s.length() - 1; ++i) {
      string palindrome = getPalindromeAround(s, i, i);
      if (palindrome.length() > longest.length())
        longest = palindrome;
      palindrome = getPalindromeAround(s, i, i + 1);
      if (palindrome.length() > longest.length())
        longest = palindrome;
    }

    return longest;
  }

private:
  string getPalindromeAround(String s, int l, int r) {
    if (s[l--] != s[r++])
      return "";

    while (l >= 0 && r <= s.length() - 1) {
      if (s[l] != s[r])
        break;
      else --l, ++r;
    }

    return s.substr(l + 1, r - l - 1);
  }

};

File: 005-longest palindromic substring/Solution005.java

public class Solution005 {
    //Attempt1: 254ms
    public String longestPalindrome(String s) {
        if (s == null || s.length() == 0)
            return "";

        String longest = s.substring(0, 1);
        for (int i = 0; i < s.length() - 1; ++i) {
            String palindrome = getPalindromeAround(s, i, i);
            if (palindrome.length() > longest.length())
                longest = palindrome;
            palindrome = getPalindromeAround(s, i, i + 1);
            if (palindrome.length() > longest.length())
                longest = palindrome;
        }

        return longest;
    }

    private String getPalindromeAround(String s, int l, int r) {
        if (s.charAt(l--) != s.charAt(r++))
            return "";

        while (l >= 0 && r <= s.length() - 1) {
            if (s.charAt(l) != s.charAt(r))
                break;
            else {
                --l; ++r;
            }
        }

        return s.substring(l + 1, r);
    }

    public static void main(String[] args) {
        String s = "bb";
        System.out.println(new Solution005().longestPalindrome(s));
    }
}
